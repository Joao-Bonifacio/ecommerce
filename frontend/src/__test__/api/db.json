{
  "users": [
    {
      "id": "uuid-1",
      "name": "John Dee",
      "email": "test@test.com",
      "nickname": "john_dee",
      "password": "@Passw0rd",
      "access_token": "fake-jwt-token"
    }
  ],
  "products": [
    {
      "id": "uuid-1",
      "title": "testing 4",
      "description": "Test product description.",
      "price": 99.99,
      "slug": "product-test1",
      "featured": true,
      "owner": 1
    },
    {
      "id": "uuid-2",
      "title": "testing 4",
      "description": "Test product description.",
      "price": 99.99,
      "slug": "product-test2",
      "featured": true,
      "owner": 1
    },
    {
      "id": "uuid-3",
      "title": "testing 4",
      "description": "Test product description.",
      "price": 99.99,
      "slug": "product-test3",
      "featured": true,
      "owner": 1
    },
    {
      "id": "uuid-4",
      "title": "testing 4",
      "description": "Test product description.",
      "price": 99.99,
      "slug": "product-test4",
      "featured": true,
      "owner": 1
    },
    {
      "id": "uuid-5",
      "title": "testing 4",
      "description": "Test product description.",
      "price": 99.99,
      "slug": "product-test5",
      "featured": false,
      "owner": 1
    },
    {
      "id": "uuid-6",
      "title": "testing 4",
      "description": "Test product description.",
      "price": 99.99,
      "slug": "product-test6",
      "featured": false,
      "owner": 1
    },
    {
      "id": "uuid-7",
      "title": "testing 4",
      "description": "Test product description.",
      "price": 99.99,
      "slug": "product-test7",
      "featured": false,
      "owner": 1
    },
    {
      "id": "uuid-8",
      "title": "testing 4",
      "description": "Test product description.",
      "price": 99.99,
      "slug": "product-test8",
      "featured": false,
      "owner": 1
    },
    {
      "id": "uuid-9",
      "title": "testing 4",
      "description": "Test product description.",
      "price": 99.99,
      "slug": "product-test9",
      "featured": false,
      "owner": 1
    },
    {
      "id": "uuid-10",
      "title": "testing 4",
      "description": "Test product description.",
      "price": 99.99,
      "slug": "product-test10",
      "featured": false,
      "owner": 1
    }
  ]
}
