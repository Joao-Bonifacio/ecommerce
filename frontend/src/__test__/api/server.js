const jsonServer = require('json-server');
const multer = require('multer');
const path = require('path');
const server = jsonServer.create();
const router = jsonServer.router('db.json');
const middlewares = jsonServer.defaults();
const upload = multer({ dest: 'upload/' });

server.use(middlewares);
server.use(jsonServer.bodyParser);

server.post('/v1/session/sign-up', (req, res) => {
  const db = router.db;
  const users = db.get('users');
  const userExists = users.find({ nickname: req.body.nickname }).value();

  if (userExists) {
    return res.status(400).json({ error: 'Nickname already exists' });
  }

  const newUser = {
    id: Date.now(),
    ...req.body,
    access_token: 'fake-jwt-token-' + Date.now()
  };
  users.push(newUser).write();
  res.status(201).json(newUser);
});

server.post('/v1/session/sign-in', (req, res) => {
  const db = router.db;
  const user = db.get('users').find({ nickname: req.body.nickname, password: req.body.password }).value();

  if (!user) {
    return res.status(401).json({ error: 'Invalid nickname or password' });
  }

  res.json({ access_token: user.access_token });
});

server.use((req, res, next) => {
  if (['/v1/products', '/v1/products/featured', '/v1/products/slug', '/v1/products/search', '/v1/products/remove', '/v1/session'].some(path => req.path.startsWith(path))) {
    if (req.method !== 'GET') {
      const auth = req.headers.authorization;
      if (!auth || !auth.startsWith('Bearer ')) {
        return res.status(401).json({ error: 'Unauthorized' });
      }
    }
  }
  next();
});

server.post('/v1/products', upload.single('file'), (req, res) => {
  const db = router.db;
  const products = db.get('products');

  const { title, description, price, slug } = req.body;

  if (products.find({ slug }).value()) {
    return res.status(400).json({ error: 'Slug already exists' });
  }

  const newProduct = {
    id: Date.now(),
    title,
    description,
    price: parseFloat(price),
    slug,
    featured: false,
    owner: 1,
    image: req.file ? `/upload/${req.file.filename}` : null
  };

  products.push(newProduct).write();

  res.status(201).json(newProduct);
});

server.patch('/v1/products/featured/:id', (req, res) => {
  const db = router.db;
  const id = parseInt(req.params.id);

  const product = db.get('products').find({ id }).value();
  if (!product) {
    return res.status(404).json({ error: 'Product not found' });
  }

  db.get('products').find({ id }).assign({ featured: true }).write();

  res.json({ message: 'Product featured' });
});

server.get('/v1/products/slug/:slug', (req, res) => {
  const db = router.db;
  const product = db.get('products').find({ slug: req.params.slug }).value();

  if (!product) {
    return res.status(404).json({ error: 'Product not found' });
  }

  res.json(product);
});

server.get('/v1/products/featured', (req, res) => {
  const db = router.db;
  const products = db.get('products').filter({ featured: true }).value();
  res.json(products);
});

server.get('/v1/products/search', (req, res) => {
  const db = router.db;
  const q = req.query.q?.toLowerCase() || '';
  const products = db.get('products').filter(p => p.title.toLowerCase().includes(q)).value();
  res.json(products);
});

server.delete('/v1/products/remove/:title', (req, res) => {
  const db = router.db;
  const title = req.params.title;
  const product = db.get('products').find({ title }).value();

  if (!product) {
    return res.status(404).json({ error: 'Product not found' });
  }

  db.get('products').remove({ title }).write();

  res.json({ message: 'Product removed' });
});

server.delete('/v1/session', (req, res) => {
  res.json({ message: 'User logged out' });
});

server.use('/upload', jsonServer.static(path.join(__dirname, 'upload')));
server.use('/v1', router);

const PORT = 8080;
server.listen(PORT, () => {
  console.log(`JSON Server running at http://localhost:${PORT}/v1`);
});
