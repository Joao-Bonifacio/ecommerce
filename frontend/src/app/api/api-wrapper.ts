import { env } from '@/env'
import { RequestInit } from 'next/dist/server/web/spec-extension/request'

export const api = (path: string, init?: RequestInit) => {
  const baseUrl = env.NEXT_PUBLIC_API_BASE_URL
  const url = new URL(baseUrl)

  url.pathname = `/v1${path.startsWith('/') ? path : `/${path}`}`

  return fetch(url.toString(), init)
}
