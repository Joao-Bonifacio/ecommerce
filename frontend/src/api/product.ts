'use server'
import { api } from '@/api/api-wrapper'
import type { Product } from './validation/types/product'
import { cookies } from 'next/headers'
import { revalidateTag } from 'next/cache'
import { validate } from './validation/zod-validate'
import {
  editProductFormSchema,
  productFormSchema,
} from './validation/product-validate'

export const getProduct = async (slug: string): Promise<Product | null> => {
  const product = await api(`/products/slug/${slug}`, {
    method: 'GET',
    next: {
      revalidate: 60 * 60, // 1 hour
    },
  }).then((data) => data.json())
  if (!product) throw new Error('Cannot find product')
  return product
}

export const getFeaturedProducts = async (): Promise<Product[] | null> => {
  const products = await api('/products/featured', {
    method: 'GET',
    next: {
      revalidate: 60 * 60, // 1 hour
    },
  }).then((data) => data.json())
  if (!Array.isArray(products)) return null
  return products
}

export const searchProducts = async (
  query: string,
): Promise<Product[] | null> => {
  const products = await api(`/products/search?q=${query}`, {
    method: 'GET',
    next: {
      revalidate: 60 * 60, // 1 hour
    },
  }).then((data) => data.json())

  if (!Array.isArray(products)) return null

  return products
}

// seller

export const getMyProducts = async (): Promise<Product[]> => {
  const token = (await cookies()).get('access_token')
  if (!token) throw new Error('Unauthorized')

  const products = await api('/seller/my-products', {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token}`,
    },
    next: {
      revalidate: 60 * 60, // 1 hour
    },
  }).then((data) => data.json())

  return products
}

export const createProduct = async (data: FormData): Promise<void> => {
  const token = (await cookies()).get('access_token')
  const upload = {
    title: data.get('title'),
    description: data.get('description'),
    price: data.get('price'),
    fileName: data.get('fileName'),
    image: data.get('image'),
    stock: data.get('stock'),
  }
  const validation = validate(upload, productFormSchema)
  if (!token || !validation) throw new Error('Unauthorized')

  const response = await api('/seller/upload', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(upload),
    next: {
      tags: ['products'],
    },
  })

  if (!response.ok) {
    throw new Error('Failed to create product')
  }

  revalidateTag('products')
}

export const editProduct = async (
  id: string,
  data: FormData,
): Promise<void> => {
  const token = (await cookies()).get('access_token')
  const upload = {
    title: data.get('title'),
    description: data.get('description'),
    price: data.get('price'),
    fileName: data.get('fileName'),
    image: data.get('image'),
    stock: data.get('stock'),
  }
  const validation = validate(upload, editProductFormSchema)
  if (!token || !validation) throw new Error('Unauthorized')

  const response = await api(`/seller/edit/${id}`, {
    method: 'PATCH',
    headers: {
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(upload),
    next: {
      tags: ['products'],
    },
  })

  if (!response.ok) {
    throw new Error('Failed to edit product')
  }
  revalidateTag('products')
}

export const featureProduct = async (id: string): Promise<void> => {
  const token = (await cookies()).get('access_token')
  if (!token) throw new Error('Unauthorized')

  const response = await api(`/seller/feature/${id}`, {
    method: 'PATCH',
    headers: {
      Authorization: `Bearer ${token}`,
    },
    next: {
      tags: ['products'],
    },
  })

  if (!response.ok) {
    throw new Error('Failed to feature product')
  }
  revalidateTag('products')
}

export const removeProduct = async (id: string): Promise<void> => {
  const token = (await cookies()).get('access_token')
  if (!token) throw new Error('Unauthorized')

  const response = await api(`/seller/remove/${id}`, {
    method: 'DELETE',
    headers: {
      Authorization: `Bearer ${token}`,
    },
    next: {
      tags: ['products'],
    },
  })

  if (!response.ok) {
    throw new Error('Failed to remove product')
  }
  revalidateTag('products')
}
