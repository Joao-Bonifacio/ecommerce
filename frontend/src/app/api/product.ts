'use server'
import { api } from '@/app/api/api-wrapper'
import type { Product } from './validation/types/product'
import { env } from '@/env'

export const getProduct = async (slug: string): Promise<Product | null> => {
  const product = await api(`/products/slug/${slug}`, {
    method: 'GET',
    next: {
      revalidate: 60 * 60, // 1 hour
    },
  }).then((data) => data.json())
  if (!product) throw new Error('Cannot find product')
  return product
}

export const getFeaturedProducts = async (): Promise<Product[] | null> => {
  const products = await api('/products/featured', {
    method: 'GET',
    next: {
      revalidate: 60 * 60, // 1 hour
    },
  }).then((data) => data.json())
  if (!products) return null
  return products
}

export const searchProducts = async (
  query: string,
): Promise<Product[] | null> => {
  const url = new URL('/v1/products/search', env.NEXT_PUBLIC_API_BASE_URL)
  url.searchParams.set('q', query)

  const products = await fetch(url, {
    method: 'GET',
    next: {
      revalidate: 60 * 60, // 1 hour
    },
  }).then((data) => data.json())

  if (!Array.isArray(products)) return null

  return products
}
