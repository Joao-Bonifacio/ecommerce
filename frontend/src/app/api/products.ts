'use server'
import { api } from '@/app/api/api-wrapper'
import type { Product } from './validation/types/product'

export const getProduct = async (slug: string): Promise<Product | null> => {
  const product = await api(`/products/${slug}`, {
    method: 'GET',
    next: {
      revalidate: 60 * 60, // 1 hour
    },
  }).then((data) => data.json())
  if (!product) throw new Error('Cannot find product')
  return product
}

export const getFeaturedProducts = async (): Promise<Product[] | null> => {
  const products = await api('/products/featured', {
    method: 'GET',
    next: {
      revalidate: 60 * 60, // 1 hour
    },
  }).then((data) => data.json())
  if (!products) return null
  return products
}

export const searchProducts = async (
  query: string,
): Promise<Product[] | null> => {
  const products = await api(`/products/search?q=${query}`, {
    method: 'GET',
    next: {
      revalidate: 60 * 60, // 1 hour
    },
  }).then((data) => data.json())
  if (!products) return null
  return products
}
