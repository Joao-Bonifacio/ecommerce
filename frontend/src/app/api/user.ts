/* eslint-disable camelcase */
'use server'

import { api } from '@/app/api/api-wrapper'
import { validate } from './validation/zod-validate'
import { signInSchema, signUpSchema } from './validation/session-validate'
import { cookies } from 'next/headers'
import { redirect } from 'next/navigation'
import { userSchema } from './validation/user.validate'
import { useUserStore } from '@/context/stores/user-store'
import { User } from './validation/types/user'

const setUserInStore = (user: User) => {
  const parsed = userSchema.safeParse(user)
  if (parsed.success) {
    useUserStore.getState().setUser(parsed.data)
  } else {
    console.error('Invalid user data from API', parsed.error)
  }
}

export const signUp = async (data: FormData): Promise<unknown | null> => {
  if (data.get('password') !== data.get('confirm-password')) {
    return null
  }
  const upload = {
    name: data.get('name'),
    email: data.get('email'),
    password: data.get('password'),
  }
  const validation = validate(upload, signUpSchema)
  if (!validation.success) return validation

  const { access_token, user } = await api('/session/sign-up', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(upload),
  }).then((data) => data.json())
  if (!access_token) return null

  const cookie = await cookies()
  cookie.set('access_token', access_token)

  setUserInStore(user)

  return redirect('/')
}

export const signIn = async (data: FormData): Promise<unknown> => {
  const upload = data.get('email')
    ? { email: data.get('email'), password: data.get('password') }
    : { nickname: data.get('nickname'), password: data.get('password') }
  const validation = validate(upload, signInSchema)
  if (!validation.success) return validation

  const { access_token, user } = await api('/session/sign-in', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(upload),
  }).then((data) => data.json())
  if (!access_token) throw new Error('Failed to sign-in')

  const cookie = await cookies()
  cookie.set('access_token', access_token)

  setUserInStore(user)

  return redirect('/')
}
