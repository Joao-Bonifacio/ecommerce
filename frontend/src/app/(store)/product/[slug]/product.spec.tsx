import { expect, vi } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react'
import ProductPage, {
  generateStaticParams,
  generateMetadata,
} from './page'
import * as productApi from '@/app/api/product'
import { Product } from '@/app/api/validation/types/product'

vi.mock('@/components/add-to-cart-button', () => ({
  __esModule: true,
  default: ({ productId }: { productId: string }) => (
    <button>Add to Cart ({productId})</button>
  ),
}))
vi.mock('@/app/api/products')

describe('Product Page', () => {
  const mockProduct: Product = {
    id: crypto.randomUUID(),
    slug: 'product-slug',
    title: 'Mock Product',
    description: 'Mock Description',
    price: 120,
    image: '/mock.jpg',
    owner: '',
    sales: 0,
    featured: false,
    ratings: [],
  }

  const mockProducts: Product[] = [
    { ...mockProduct },
    {
      ...mockProduct,
      id: crypto.randomUUID(),
      slug: 'other-product',
      title: 'Other Product',
    },
  ]

  beforeEach(() => {
    vi.resetAllMocks()
  })

  it('should render product details', async () => {
    vi.spyOn(productApi, 'getProduct').mockResolvedValue(mockProduct)

    render(await ProductPage({ params: { slug: 'product-slug' } }))

    await waitFor(() => {
      expect(screen.getByText('Mock Product')).toBeInTheDocument()
      expect(screen.getByText('Mock Description')).toBeInTheDocument()
      expect(screen.getByText('$120.00')).toBeInTheDocument()
      expect(screen.getByText('In 12x of $10.00')).toBeInTheDocument()
      expect(screen.getByText(/Add to Cart/)).toBeInTheDocument()
    })
  })

  it('should return null if product not found', async () => {
    vi.spyOn(productApi, 'getProduct').mockResolvedValue(null)

    const { container } = render(
      await ProductPage({ params: { slug: 'nonexistent-slug' } })
    )
    expect(container.firstChild).toBeNull()
  })

  it('generateStaticParams should return product slugs', async () => {
    vi.spyOn(productApi, 'getFeaturedProducts').mockResolvedValue(mockProducts)

    const result = await generateStaticParams()
    expect(result).toEqual([
      { slug: 'product-slug' },
      { slug: 'other-product' },
    ])
  })

  it('generateStaticParams should return null when no products', async () => {
    vi.spyOn(productApi, 'getFeaturedProducts').mockResolvedValue(null)

    const result = await generateStaticParams()
    expect(result).toBeNull()
  })

  it('generateMetadata should return title if product found', async () => {
    vi.spyOn(productApi, 'getProduct').mockResolvedValue(mockProduct)

    const metadata = await generateMetadata({ params: { slug: 'product-slug' } })
    expect(metadata).toEqual({ title: 'Mock Product' })
  })

  it('generateMetadata should return empty object if product not found', async () => {
    vi.spyOn(productApi, 'getProduct').mockResolvedValue(null)

    const metadata = await generateMetadata({ params: { slug: 'invalid' } })
    expect(metadata).toEqual({})
  })
})
