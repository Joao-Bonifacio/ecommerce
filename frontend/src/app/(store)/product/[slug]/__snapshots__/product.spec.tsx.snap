// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductPage > ProductPage component > renders product details when product found 1`] = `
<div>
  <div
    class="relative grid max-h-[860px] grid-cols-3"
  >
    <div
      class="col-span-2 overflow-hidden"
    >
      <img
        alt="This is a mock product"
        class="rounded-r-xl"
        data-nimg="1"
        decoding="async"
        height="1000"
        loading="lazy"
        src="/_next/image?url=%2Fmock.jpg&w=2048&q=100"
        srcset="/_next/image?url=%2Fmock.jpg&w=1080&q=100 1x, /_next/image?url=%2Fmock.jpg&w=2048&q=100 2x"
        style="color: transparent;"
        width="1000"
      />
    </div>
    <div
      class="flex flex-col justify-center px-12"
    >
      <h1
        class="!text-4xl font-bold leading-tight !py-2 !mb-5"
      >
        Mock Product
      </h1>
      <p
        class="mt-2 leading-relaxed text-zinc-400 !text-lg !p-y2 !mb-5"
      >
        This is a mock product
      </p>
      <div
        class="mt-8 flex items-center gap-3 !text-md"
      >
        <span
          class="inline-block rounded-full bg-violet-500 !p-3 font-semibold"
        >
          $120.00
        </span>
        <span
          class="text-zinc-400 !text-lg"
        >
          In 12x of
           
          $10.00
        </span>
      </div>
      <button>
        Add to cart mock
      </button>
    </div>
    <div
      class="h-5 w-full"
    />
  </div>
</div>
`;
